llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -8061,10 +8062,20 @@ SDValue SelectionDAG::getMemBasePlusOffset(SDValue Base, TypeSize Offset,
 
 SDValue SelectionDAG::getMemBasePlusOffset(SDValue Ptr, SDValue Offset,
                                            const SDLoc &DL,
-                                           const SDNodeFlags Flags) {
+                                           const SDNodeFlags Flags,
+                                           const bool Inverted) {
   assert(Offset.getValueType().isInteger());
   EVT BasePtrVT = Ptr.getValueType();
-  return getNode(ISD::ADD, DL, BasePtrVT, Ptr, Offset, Flags);
+  if (!this->getTarget().shouldPreservePtrArith(
```

Comment:
For CHERI it might be good to pass the BasePtrVT to shouldPreservePtrArith() as well, but I guess that change can be made later. So no need to fix this in this PR.